File: dnSpy/dnSpy/Roslyn/Roslyn.ExpressionCompiler/CSharp/CSharpExpressionCompiler/GeneratedNames2.cs
using System.Diagnostics.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp.Symbols;
using Microsoft.CodeAnalysis.ExpressionEvaluator;

namespace Microsoft.CodeAnalysis.CSharp.ExpressionEvaluator
{
    static class GeneratedNames2
    {
        public static GeneratedNameKind GetKind(this CompilerKind compiler, string name)
        {
            return CommonGeneratedNames.GetKind(compiler, name).ToGeneratedNameKind();
        }

        public static bool IsSynthesizedLocalName(this CompilerKind compiler, string name)
        {
            return CommonGeneratedNames.IsSynthesizedLocalName(compiler, name);
        }

        public static bool TryParseSlotIndex(this CompilerKind compiler, string fieldName, out int slotIndex)
        {
            return CommonGeneratedNames.TryParseSlotIndex(compiler, fieldName, out slotIndex);
        }

        public static bool TryParseSourceMethodNameFromGeneratedName(this CompilerKind compiler, string generatedName, GeneratedNameKind requiredKind, [NotNullWhen(true)] out string? methodName)
        {
            return CommonGeneratedNames.TryParseSourceMethodNameFromGeneratedName(compiler, generatedName, requiredKind.ToCommonGeneratedNameKind(), out methodName);
        }

        public static bool TryParseGeneratedName(this CompilerKind compiler, string name, out GeneratedNameKind kind, [NotNullWhen(true)] out string? part)
        {
            var res = CommonGeneratedNames.TryParseGeneratedName(compiler, name, out var commonKind, out part);
            kind = commonKind.ToGeneratedNameKind();
            return res;
        }

        public static string? GetUnmangledTypeParameterName(this CompilerKind compiler, string typeParameterName)
        {
            return CommonGeneratedNames.GetUnmangledTypeParameterName(compiler, typeParameterName);
        }

        public static bool IsDisplayClassInstance(this CompilerKind compiler, string fieldType, string fieldName)
        {
            return CommonGeneratedNames.IsDisplayClassInstance(compiler, fieldType, fieldName);
        }
    }
}

File: dnSpy/dnSpy/Roslyn/Roslyn.ExpressionCompiler/VisualBasic/BasicExpressionCompiler/GeneratedNames2.vb
Imports System.Collections.Immutable
Imports System.Runtime.CompilerServices
Imports System.Runtime.InteropServices
Imports Microsoft.CodeAnalysis.ExpressionEvaluator
Imports Microsoft.CodeAnalysis.VisualBasic.Symbols

Namespace Microsoft.CodeAnalysis.VisualBasic.ExpressionEvaluator
    Friend Module GeneratedNames2
        <Extension>
        Friend Function GetKind(compiler As CompilerKind, name As String) As GeneratedNameKind
            Return CommonGeneratedNames.GetKind(compiler, name).ToGeneratedNameKind()
        End Function

        <Extension>
        Friend Function GetKind2(compiler As CompilerKind, name As String) As CommonGeneratedNameKind
            Return CommonGeneratedNames.GetKind(compiler, name)
        End Function

        <Extension>
        Friend Function TryParseGeneratedName(compiler As CompilerKind, name As String, <Out> ByRef kind As CommonGeneratedNameKind, <Out> ByRef part As String) As Boolean
            Dim res = CommonGeneratedNames.TryParseGeneratedName(compiler, name, kind, part)
            Return res
        End Function

        <Extension>
        Friend Function TryParseHoistedUserVariableName(compiler As CompilerKind, proxyName As String, <Out> ByRef variableName As String) As Boolean
            Return CommonGeneratedNames.TryParseHoistedUserVariableName(compiler, proxyName, variableName)
        End Function

        <Extension>
        Friend Function TryParseStateMachineHoistedUserVariableOrDisplayClassName(compiler As CompilerKind, proxyName As String, <Out> ByRef variableName As String, <Out()> ByRef index As Integer) As Boolean
            Return CommonGeneratedNames.TryParseStateMachineHoistedUserVariableOrDisplayClassName(compiler, proxyName, variableName, index)
        End Function

        <Extension>
        Public Function TryParseStateMachineTypeName(compiler As CompilerKind, stateMachineTypeName As String, <Out> ByRef methodName As String) As Boolean
            Return CommonGeneratedNames.TryParseStateMachineTypeName(compiler, stateMachineTypeName, methodName)
        End Function

        <Extension>
        Friend Function TryParseStaticLocalFieldName(compiler As CompilerKind, fieldName As String, <Out> ByRef methodName As String, <Out> ByRef methodSignature As String, <Out> ByRef localName As String) As Boolean
            Return CommonGeneratedNames.TryParseStaticLocalFieldName(compiler, fieldName, methodName, methodSignature, localName)
        End Function

        <Extension>
        Public Function GetUnmangledTypeParameterName(compiler As CompilerKind, typeParameterName As String) As String
            Return CommonGeneratedNames.GetUnmangledTypeParameterName(compiler, typeParameterName)
        End Function

        <Extension>
        Public Function ContainsHoistedMeName(compiler As CompilerKind, dict As ImmutableDictionary(Of String, DisplayClassVariable)) As Boolean
            For Each kv In dict
                If compiler.GetKind(kv.Key) = GeneratedNameKind.HoistedMeField Then
                    Return True
                End If
            Next
            Return False
        End Function

        <Extension>
        Public Function ContainsHoistedMeName(compiler As CompilerKind, list As IEnumerable(Of String)) As String
            For Each name In list
                If compiler.GetKind(name) = GeneratedNameKind.HoistedMeField Then
                    Return True
                End If
            Next
            Return False
        End Function

        <Extension>
        Public Function IsDisplayClassInstance(compiler As CompilerKind, fieldType As String, fieldName As String) As String
            Return CommonGeneratedNames.IsDisplayClassInstance(compiler, fieldType, fieldName)
        End Function
    End Module
End Namespace
